### ISSUE TITLE
Early stopping during training

### ISSUE THREAD

**Issue by kinggongzilla** [pytorch/torchtune#883]
Is there a way to evaluate the model performance during training on a validation dataset and only save a new checkpoint if it achieves lower validation loss?

**Comment by rohan-varma** [comment]
Hi @kinggongzilla, thanks for filing this issue!

Currently we only support an early stopping that's based on the # of steps taken in an epoch, i.e. you can set `max_steps_per_epoch` flag in the [configuration](https://github.com/pytorch/torchtune/blob/cc2dd05f47f5de5737e8d8cc3666f8c65946c78d/recipes/configs/llama3/8B_full.yaml#L59) to early stop your model based on a # of steps.

However this doesn't satisfy your use case of only early stopping / saving a checkpoint based on some validation results. 

In training evaluation + stopping criteria based on evaluation is a large space we haven't looked deeply into, what do you folks think @ebsmothers @RdoubleA? I could see a future in which we allow users to specify a validation dataset or validation split, and incorporate validation metrics into our checkpointer for whether to save a checkpoint or not. This is definitely something we could look at enabling in the future if there's sufficient interest.

**Comment by kinggongzilla** [comment]
Thanks for the quick reply!
Being able to define a validation dataset and do early stoppingbased on the validation loss would definitely be super helpful.

**Comment by optimass** [comment]
+1 this would be super useful.

**Comment by Some-random** [comment]
+1 Would be super useful!

**Comment by ebsmothers** [comment]
Thanks all for the comments. This feature (along with general validation loops) are fairly high on our wishlist right now. We still need to do a bit of design to make sure it's not too intrusive into our recipes, but definitely hear you on the need for this feature. We will keep you posted here!

**Comment by Tandon-A** [comment]
Hi @ebsmothers, 

Is there any update on this?

**Comment by karrtikiyer** [comment]
+1 Would be super useful!
